### PR TITLE
Support globs in c-sources

### PR DESCRIPTION
**Pull request by mitchellwrosen**
#118

### CODE REVIEW

**Comment by mitchellwrosen** [comment]
This is ready for review, the CI errors are all aeson-1.0.0.0 related.

**Comment by mitchellwrosen** [comment]
Anything else I should do for this @sol?

**Comment by sol** [comment]
Sorry for being lazy on this. Can you rebase this on current master, and I will try to take care of this soon.

**Comment by mitchellwrosen** [comment]
No worries! Rebased.

**Comment by sol** [comment]
Hey Mitchell, sorry again for the delay. I am currently working on a voice programming set up, and I hope this will allow me to avoid situations like that in the future.

Regarding your code, I would try to make the code for the main feature more concise. I may try some things here myself.

But some things you did are useful on their own. Let's split them out into a separate PR.

File: src/Hpack/Config.hs [hunk 1]
```diff
@@ -444,25 +444,27 @@ mkPackage :: FilePath -> (CaptureUnknownFields (Section PackageConfig)) -> IO ([
...
     ++ concatMap sectionSourceDirs benchmarks
     )
 
   (extraSourceFilesWarnings, extraSourceFiles) <-
-    expandGlobs dir (fromMaybeList packageConfigExtraSourceFiles)
+    expandGlobs "extra-source-files" dir (fromMaybeList packageConfigExtraSourceFiles)
 
   (dataFilesWarnings, dataFiles) <-
-    expandGlobs dir (fromMaybeList packageConfigDataFiles)
+    expandGlobs "data-files" dir (fromMaybeList packageConfigDataFiles)
```
Comment (sol) [inline src/Hpack/Config.hs]: Can you move this into a separate commit?

File: src/Hpack/Util.hs [hunk 2]
```diff
@@ -101,14 +101,14 @@ tryReadFile file = do
...
-expandGlobs dir patterns = do
+expandGlobs :: String -> FilePath -> [String] -> IO ([String], [FilePath])
+expandGlobs name dir patterns = do
   files <- (fst <$> globDir compiledPatterns dir) >>= mapM removeDirectories
   let warnings = [warn pattern | ([], pattern) <- zip files patterns]
   return (warnings, combineResults files)
   where
     combineResults = nub . sort . map (toPosixFilePath . makeRelative dir) . concat
-    warn pattern = "Specified pattern " ++ show pattern ++ " for extra-source-files does not match any files"
+    warn pattern = "Specified pattern " ++ show pattern ++ " for " ++ name ++ " does not match any files"
```
Comment (sol) [inline src/Hpack/Util.hs]: this too

File: test/Hpack/UtilSpec.hs [hunk 3]
```diff
@@ -97,22 +97,22 @@ spec = do
   describe "expandGlobs" $ around withTempDirectory $ do
     it "accepts simple files" $ \dir -> do
         touch (dir </> "foo.js")
-        expandGlobs dir ["foo.js"] `shouldReturn` ([], ["foo.js"])
+        expandGlobs "" dir ["foo.js"] `shouldReturn` ([], ["foo.js"])
```
Comment (sol) [inline test/Hpack/UtilSpec.hs]: I think for the purpose of documentation e.g. `"field-name"` is more suitable than `""` or `"XXX"`

File: src/Hpack/Config.hs [hunk 4]
```diff
@@ -566,12 +573,23 @@ mkPackage dir (CaptureUnknownFields unknownFields globalOptions@Section{sectionD
       where
         fromGithub = (++ "/issues") . sourceRepositoryUrl <$> github
 
-toLibrary :: FilePath -> String -> Section global -> Section LibrarySection -> IO (Section Library)
-toLibrary dir name globalOptions library = traverse fromLibrarySection sect
+-- | Expand the @c-sources@ globs in a 'Section' and return the modified
+-- 'Section' along with any warnings emitted.
+globSectionCSources :: FilePath -> Section a -> IO ([String], Section a)
+globSectionCSources dir sect = do
```
Comment (sol) [inline src/Hpack/Config.hs]: I would remove the comment and rename the function to something like `expandCSources`

File: src/Hpack/Config.hs [hunk 5]
```diff
@@ -566,12 +573,23 @@ mkPackage dir (CaptureUnknownFields unknownFields globalOptions@Section{sectionD
...
+  (cSourcesWarnings, cSourcesFiles) <-
```
Comment (sol) [inline src/Hpack/Config.hs]: If only it's more convenient for voice programming, I would prefer `(warnings, files)`

**Comment by sol** [comment]
I applied some refactoring on top of your changes. To see the total diff you may want to look at #131.
